### PR TITLE
Feature lb ondemand caching

### PR DESCRIPTION
**Pull request by derekolk**

### CODE REVIEW

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  }
+
+  /**
+   * Generically will move cached data from onDemand to resource (Server Group, Load Balancer, Security Group) namespace.
+   * @param objectMapper
+   * @param cacheResultBuilder
+   * @param serverGroupKey
+   */
+  void moveOnDemandDataToNamespace(ObjectMapper objectMapper, TypeReference<Map<String, List<MutableCacheData>>> typeReference, CacheResultBuilder cacheResultBuilder, String key) {
+    String cacheResults = cacheResultBuilder.onDemand.toKeep[key].attributes.cacheResults as String
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy:55]: Why the `as String`? What is `cacheResults`?

Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy:55]: Attributes is a Map<String, Object> that contains a key/value pair of cacheResults and a String that contains a json representation 'cacheResults': '{ json representation }'  ... Type casting and object to String.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+   * Builds on demand cache result based upon the cache result.
+   * @param onDemandAgentType
+   * @param cacheResult
+   * @param namespace
+   * @param key
+   * @param providerCache
+   * @return
+   */
+  OnDemandResult buildOnDemandCache(Object cachedItem
+                                    , String onDemandAgentType
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy:102]: That is an odd format

**Comment by varikin** [comment]
LGTM, at least until I try consuming it :)

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    cacheData ? cacheData.attributes.cacheTime >= cacheResultBuilder.startTime : false
+  }
+
+  /**
+   * Generically will move cached data from onDemand to resource (Server Group, Load Balancer, Security Group) namespace.
+   * @param objectMapper
+   * @param cacheResultBuilder
+   * @param serverGroupKey
+   */
+  void moveOnDemandDataToNamespace(ObjectMapper objectMapper, TypeReference<Map<String, List<MutableCacheData>>> typeReference, CacheResultBuilder cacheResultBuilder, String key) {
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy:54]: This seems a bit confusing - you are mutating the parameters passed to this method. Might be more clear to return the cacheResultBuilder for clarity.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+   */
+  OnDemandResult buildOnDemandCache(Object cachedItem
+                                    , String onDemandAgentType
+                                    , CacheResult cacheResult
+                                    , String namespace
+                                    , String key) {
+    OnDemandResult result = new OnDemandResult(sourceAgentType: onDemandAgentType, cacheResult: cacheResult, evictions: [:].withDefault { _ -> [] })
+
+    if (!cachedItem) {
+      // Evict this load balancer if it no longer exists.
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy]: Change comment, as this is a generic trait

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackNetworkingV2Provider.groovy [hunk 5]
```diff
@@ -50,13 +50,26 @@ class OpenstackNetworkingV2Provider implements OpenstackNetworkingProvider, Open
...
   List<? extends LbPool> getAllLoadBalancerPools(final String region) {
+    getInternalAllLoadBalancerPools(region)
+  }
+
+  List<? extends LbPool> getInternalAllLoadBalancerPools(final String region, Map<String, String> filterParameters = null) {
     List<? extends LbPool> pools = handleRequest {
-      getRegionClient(region).networking().loadbalancers().lbPool().list()
+      getRegionClient(region).networking().loadbalancers().lbPool().list(filterParameters)
     }
     pools
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackNetworkingV2Provider.groovy]: Just return handleRequest per conventions

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/AbstractOpenstackCachingAgent.groovy [hunk 6]
```diff
@@ -34,6 +39,7 @@ abstract class AbstractOpenstackCachingAgent implements CachingAgent, AccountAwa
   AbstractOpenstackCachingAgent(OpenstackNamedAccountCredentials account, String region) {
     this.account = account
     this.region = region
+    this.typeReference = new TypeReference<Map<String, List<MutableCacheData>>>() {}
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/AbstractOpenstackCachingAgent.groovy]: Why is this hardcoded here? You could just hardcode the field name instead.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy [hunk 7]
```diff
@@ -45,80 +53,149 @@ import static com.netflix.spinnaker.clouddriver.openstack.cache.Keys.Namespace.V
...
+    OnDemandResult result = null
+
+    if (data.containsKey("loadBalancerName") && data.account == accountName && data.region == region) {
+      String loadBalancerName = data.loadBalancerName.toString()
+
+      LbPool pool = metricsSupport.readData {
+        try {
+          clientProvider.getLoadBalancerPoolByName(region, loadBalancerName)
+        } catch (OpenstackProviderException ope) {
+          //Do nothing ... Exception is thrown if a stack isn't found
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy]: change from stack to load balancer

**Comment by drmaas** [comment]
lgtm after comments are addressed.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy (hunk 2 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy:102]: fixed.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy (hunk 4 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy]: done.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackNetworkingV2Provider.groovy (hunk 5 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackNetworkingV2Provider.groovy]: done.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/AbstractOpenstackCachingAgent.groovy (hunk 6 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/AbstractOpenstackCachingAgent.groovy]: Changed to assign at instance level.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy (hunk 7 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy]: Done.

**Pull request closed without merging**